### PR TITLE
Configuration support for both ini and json files. 

### PR DESCRIPTION
**Pull request by mgarewal-r7**
Logentries log following can now be configured utilizing a json configuration file or an ini configuration file.

The default file used is logging.json, if that file is not present, the system will look for a relevant ini config file to use.

The system will check if the configuration files are present in the local directory (ordinary user log in) or the root directory (/etc/le/).

### CODE REVIEW

**Review by jsouthern-r7 — APPROVED** [review]
It's so good to see that an initial set of unit tests has been added. Nice work, Mandy.

File: src/TestJsonConfig.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import unittest
+import json
+import metrics
+from config import Config
+
+CONFIG = Config()
```
Comment (jyeung-r7) [inline src/TestJsonConfig.py]: Avoid global for unit-tests

File: src/__init__.py [hunk 2]
```diff
@@ -1,2 +1,2 @@
-"""Le Agent Version"""
+
```
Comment (jyeung-r7) [inline src/__init__.py]: What's the reason to remove this?

File: src/config.py [hunk 3]
```diff
@@ -12,15 +12,17 @@
 
 import metrics
 import utils
+import json
 from log import log
 from configured_log import ConfiguredLog
 from constants import NOT_SET, EXIT_OK, MULTILOG_USAGE, DESTINATION_PARAM, TOKEN_PARAM
-
+import glob
```
Comment (jyeung-r7) [inline src/config.py]: Please move this above all the "from" imports

File: src/config.py [hunk 4]
```diff
@@ -12,15 +12,17 @@
...
 
 DEFAULT_USER_KEY = NOT_SET
 DEFAULT_AGENT_KEY = NOT_SET
 PID_FILE = '/var/run/logentries.pid'
-LE_CONFIG = 'config'  # Default configuration file
+LE_CONFIG = 'config'
+JSON_CONFIG = 'logging.json' # Default configuration file
```
Comment (jyeung-r7) [inline src/config.py]: Is this going to be the default?

File: src/config.py [hunk 5]
```diff
@@ -301,6 +307,67 @@ def load(self, load_include_dirs=True):
         """
 
         try:
+            PULL_SERVER_SIDE_CONFIG_PARAM = 'True'
```
Comment (jyeung-r7) [inline src/config.py]: Why is this set to true? When the default is 
PULL_SERVER_SIDE_CONFIG_PARAM = 'pull-server-side-config'

File: src/config.py [hunk 6]
```diff
@@ -301,6 +307,67 @@ def load(self, load_include_dirs=True):
...
+            # Read configuration files from default directories
+            config_files = [self.config_filename]
+            if load_include_dirs:
+                config_files.extend(self._list_configs(self.config_d))
+            self._set_config_file_perms(config_files)
+
+            # Read in json config file
+            with open(config_files[0]) as data_file:
```
Comment (jyeung-r7) [inline src/config.py]: Why does it load the first file? config_files[0]?

File: src/config.py [hunk 7]
```diff
@@ -301,6 +307,67 @@ def load(self, load_include_dirs=True):
...
+                with open(self._list_configs(self.include)) as config_data:
+                    config_files.extend(json.loads(config_data).read())
+
+            log.log.debug('Configuration files loaded: %s', ', '.join(config_files))
+            self._load_parameters_json(d_configFile)
+            self._configure_proxy_json(d_configFile)
+            new_suppress_ssl = d_configFile.get(SUPPRESS_SSL_PARAM)
+
+            if new_suppress_ssl == 'True':
+                self.suppress_ssl = new_suppress_ssl == 'True'
```
Comment (jyeung-r7) [inline src/config.py]: Why is this comparing new_suppress_ssl ('True') to 'True'

File: src/le.py [hunk 8]
```diff
@@ -1235,8 +1238,11 @@ def config_formatters():
 
 def extract_token(log_):
     """Extract the log token value if it exists"""
-    return utils.safe_get(log_, 'log', 'tokens')[0]
-
+    try:
```
Comment (jyeung-r7) [inline src/le.py]: What's the reason to replace utils.safe_get ?

File: src/le.py [hunk 9]
```diff
@@ -1298,7 +1304,8 @@ def start_followers(default_transport, states):
         log_name = log_['log']['name']
         log_id = log_['log']['id']
         log_token = extract_token(log_)
-        #TODO check for no log token
+        if log_token is None:
```
Comment (jyeung-r7) [inline src/le.py]: What's the reason to set it to "" ?

File: src/utils.py [hunk 10]
```diff
@@ -533,7 +534,7 @@ def print_total(elems, name):
     elif total == 1:
         report("1 " + name)
     else:
-        report("%d %ss" % (total, name))
+        report("%d %ss \n" % (total, name))
```
Comment (jyeung-r7) [inline src/utils.py]: is the 's' necessary?

File: test/tests.d/bw_server_side_http_put.sh [hunk 11]
```diff
@@ -14,8 +14,9 @@ $LE init --account-key=$ACCOUNT_KEY
 #e Initialized
 
 cat $CONFIG
-#o [Main]
-#o user-key = f720fe54-879a-11e4-81ac-277d856f873e
+#o [Main]879a-11e4-81ac-277d856f873e
```
Comment (jyeung-r7) [inline test/tests.d/bw_server_side_http_put.sh]: Looks corrupted.

File: test/tests.d/bw_server_side_http_put.sh [hunk 12]
```diff
@@ -30,20 +31,12 @@ cat $CONFIG
 Testcase 'Register'
 
 $LE register --name Name --hostname Hostname
-#e Configuration files loaded: sandbox_config
```
Comment (jyeung-r7) [inline test/tests.d/bw_server_side_http_put.sh]: Whats the reason for this removal?

File: src/TestJsonConfig.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        expected_dict = {'net': 'sum eth0', 'space' : '/', 'disk': 'sum sda4 sda5', 'interval': '60s', 'swap': 'system',
+                         'vcpu': 'core', 'cpu': 'system', 'mem': 'system', 'token': None}
+
+        result_dict = self.metrics.load_json(self.d_configFile)
+        print('metrics loaded are: ', result_dict)
+
+        self.assertDictEqual(result_dict, expected_dict, msg=None)
+
+    # test the _load_configured_logs_json() method correctly pulls the right parameters and associated values.
+    def test_load_configured_logs_json(self):
```
Comment (jyeung-r7) [inline src/TestJsonConfig.py]: Let's add more unit-test. 
What about configuration without name? Without token? Without path?
Will it warn or error out if any is missing or if the config is invalid?

File: src/config.py (hunk 4 above)
Comment (mgarewal-r7) [reply src/config.py]: Yes.

File: src/config.py (hunk 6 above)
Comment (mgarewal-r7) [reply src/config.py]: When load_json is called, it has already been established that a json config file exists - see load() method.
In this case we only have one json config file, so even though config_files extends the files in the directory, we only need the first element of the list which is the json config file by default (see config_filename).

File: src/config.py (hunk 7 above)
Comment (mgarewal-r7) [reply src/config.py]: Logentries code logic as follows:
Its comparing whether or not new_suppress_ssl is string 'True'. 
If it is set self.suppress_ssl to boolean True. Otherwise is it kept at its default value of boolean False.

File: src/le.py (hunk 8 above)
Comment (mgarewal-r7) [reply src/le.py]: It doesnt work for ini files or json files. This is the stack trace if changed this way.
  File "/home/mgarewal/rapid7/logentries-json/src/le.py", line 2183, in <module>
    main()
  File "/home/mgarewal/rapid7/logentries-json/src/le.py", line 2175, in main
    main_root()
  File "/home/mgarewal/rapid7/logentries-json/src/le.py", line 2168, in main_root
    return func(args[1:])
  File "/home/mgarewal/rapid7/logentries-json/src/le.py", line 1532, in cmd_monitor
    start_followers(default_transport, state)
  File "/home/mgarewal/rapid7/logentries-json/src/le.py", line 1288, in start_followers
    log_token = extract_token(log_)
  File "/home/mgarewal/rapid7/logentries-json/src/le.py", line 1227, in extract_token
    return utils.safe_get(log_, 'log', 'tokens')[0]
IndexError: list index out of range

File: src/le.py (hunk 9 above)
Comment (mgarewal-r7) [reply src/le.py]: Im not sure, better ask Emma on this one. I didnt make this change.

File: src/utils.py (hunk 10 above)
Comment (mgarewal-r7) [reply src/utils.py]: Logentries code:
If you see what is being printed. It just represents the name as plural.

File: test/tests.d/bw_server_side_http_put.sh (hunk 11 above)
Comment (mgarewal-r7) [reply test/tests.d/bw_server_side_http_put.sh]: Better ask Emma on this one. I didnt make this change.

File: test/tests.d/bw_server_side_http_put.sh (hunk 12 above)
Comment (mgarewal-r7) [reply test/tests.d/bw_server_side_http_put.sh]: Better ask Emma on this one. I didnt make this change.

File: src/TestJsonConfig.py (hunk 13 above)
Comment (mgarewal-r7) [reply src/TestJsonConfig.py]: It will send warning messages. Ill add more unit tests here.

Comment (mgarewal-r7) [reply src/TestJsonConfig.py]: The specific load methods for ini and json files will return False if no config file is present. I cant write a unit test for this as these methods do not accept parameters. It is built in.

File: src/config.py (hunk 7 above)
Comment (jyeung-r7) [reply src/config.py]: Based on the logic, why is it not this?
if new_suppress_ssl == 'True':
               self.suppress_ssl = True

Comment (mgarewal-r7) [reply src/config.py]: Yes, I thought so too. I guess its in keeping with the rest of the code base.
I can change their code to this if you want.

File: src/TestJsonConfig.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+          }
+        ]
+      }
+    }'''
+
+    json_file_incorrect_token = '''{
+      "config": {
+        "hostname": "mgarewal",
+        "endpoint": "data.logentries.com",
+        "user-key": "4a22b0ba-593a-41e9-9271-bf4475b09f38",
```
Comment (jyeung-r7) [inline src/TestJsonConfig.py]: Let's use a randomly generated UUID instead of the actual ID.

File: src/TestJsonConfig.py [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+            'entry_identifier' : '',
+            'destination' : ''
+          }
+
+        result_dict = CONFIG._load_configured_logs_json(self.d_configFile)
+        print('configured log parameters are: ', result_dict)
+
+        self.assertDictEqual(result_dict, expected_dict, msg=None)
+
+    # test the _load_configured_logs_json() when incorrect token entered
```
Comment (jsouthern-r7) [inline src/TestJsonConfig.py]: Is this supposed to be a test, or just a support function for another test?

File: src/TestJsonConfig.py [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+    def test_load_configured_logs_json_token(self):
```
Comment (jsouthern-r7) [inline src/TestJsonConfig.py]: Is this supposed to be a test, or just a support function for another test?

File: src/TestJsonConfig.py [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+        d_configFile = d_conf['config']
+
+        CONFIG._load_configured_logs_json(self.d_configFile)
+
+        # result matches error message
+        logging.getLogger().info("Invalid log token `%s' in application `%s'.",
+                        token, name)
+
+    # test the _load_configured_logs_json() when path does not exist.
+    def test_load_configured_logs_json_path(self):
```
Comment (jsouthern-r7) [inline src/TestJsonConfig.py]: Is this supposed to be a test, or just a support function for another test?

File: src/config.py [hunk 18]
```diff
@@ -48,21 +48,23 @@
 KEY_LEN = 36
 LE_DEFAULT_SSL_PORT = 20000
 LE_DEFAULT_NON_SSL_PORT = 10000
-
+CONFIG_PARAM = 'config'
+CONFIG_LOGS = 'logs'
+LOG_NAME = 'name'
+EMPTY = 0
```
Comment (jyeung-r7) [inline src/config.py]: What is this used for?

File: src/config.py [hunk 19]
```diff
@@ -301,6 +306,55 @@ def load(self, load_include_dirs=True):
...
+            if self.state_file == NOT_SET:
+                state_file_str = d_configFile.get(STATE_FILE_PARAM)
+                if state_file_str:
+                    self.state_file = state_file_str
+
+            self.metrics.load_json(d_configFile)
+            self._load_configured_logs_json(d_configFile, log.log)
+
+        except ValueError:
+            print('JSON load error')
```
Comment (jyeung-r7) [inline src/config.py]: Please use a logger instead of printing to stdout

File: src/le.py (hunk 8 above)
Comment (jyeung-r7) [reply src/le.py]: Have you tried this?
```
if 'log' in log_ and utils.safe_get(log_, 'log', 'souce_type') is 'token':
  return utils.safe_get(log_, 'log', 'tokens')[0]
```

File: test/test_json_config.py [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+import json
+import sys, os
+sys.path.append(os.path.dirname(os.path.realpath(__file__)) + "/../src")
+import sys
+import logging
+import mock
+from config import Config
+from unittest import main
+
+
```
Comment (jyeung-r7) [inline test/test_json_config.py]: empty space

File: logentries/le.py [hunk 21]
```diff
@@ -29,11 +29,8 @@
 import time
 import traceback
 import requests
-try:
-    import Queue as queue
-except ImportError:
-    import queue
-    from queue import Queue
+import queue
```
Comment (jyeung-r7) [inline logentries/le.py]: Why was this changed?

File: test/test_json_config.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+import sys, os
+sys.path.append(os.path.dirname(os.path.realpath(__file__)) + "/../src")
+import sys
+import logging
+import mock
+from logentries.config import Config
+import logentries.metrics as metrics
+
+
+
```
Comment (jyeung-r7) [inline test/test_json_config.py]: Is this space intentional?
